### DISCUSSION TITLE
ioc.fake in AdonisV5

### DISCUSSION THREAD

**Discussion by vowar** [adonisjs/core#1284] [Help] [answered]
Is it possible to do something like this in adonisV5 to fake services for tests?

```
ioc.fake('Twilio', () => {
      return {
        sendSMS: async () => { ... },
        sendVerificationCode: async () => { ... },
        checkVerificationCode: async () => { ... }
      }
    })
```

**Comment by thetutlage** [comment]
If you can share the exact test code, then I can try to help on how to achieve this. The current example is too abstract

**Comment by 4konung** [comment]
Hi! I have maybe the same question.
I have a controller that uses Twilio service like this:
```javascript
import twilio from '@ioc:Services/Twilio'

export default class SomeController {

public async sendSms({ auth }: HttpContextContract) {
    const user = auth.user
    await twilio.sendSms(user.phone, 'message: Some text')

    return response.apiSuccess()
  }
}
```

and I want to 'fake' this service method when tests are running.

**Reply by thetutlage** [reply]
Can you share, how exactly you are running tests? Like what's your setup?

**Reply by 4konung** [reply]
Hi. I'm using setup from [blog](https://preview.adonisjs.com/blog/running-tests-in-adonisjs-v5/).
My tests looks like this. I'm sending requests to api-endpoints to test some app flows. I want to fake 3rd party services like twilio from previous message.
```javascript
test.group('CarDataController', (group) => {
    test('Should send message to the user', async () => {
    await supertest(BASE_URL).get('/api/send-sms').expect(200)
  })
})
```

**Reply by thetutlage** [reply]
Since, you are pulling the `Twilio` service from the IoC container. You should be able to define the fake as follows:

```ts
ioc.fake('Services/Twilio', () => {
  return // fake implementation
})
```

**Reply by 4konung** [reply]
I'm new to Adonis, from where can I import "ioc" in my test files?

**Reply by thetutlage** [reply]
Ohhh. Should have share a complete example

```ts
import Application from '@ioc:Adonis/Core/Application'

Application.container.fake('Services/Twilio', () => {
  
})
```

**Reply by eidellev** [reply]
Hi @thetutlage, 
I'm new to adonis  and enjoying it so far. Thanks for your hard work :-)

`.fake` doesn't seem to do anything, but if I replace it with `singleton` it overrides the original.
This is my test:
```typescript
import test from 'japa';
import Application from '@ioc:Adonis/Core/Application';
import supertest from 'supertest';

test.group('Git Controller', () => {
  test('Check out default branch', async (assert) => {
    Application.container.fake('Services/Git', () => {
      return {
        checkout() {
          doTheThing();
        },
      };
    });

    const response = await supertest(BASE_URL)
      .post('/git/login')
      .send({
        url: 'https://www.sapiens.com/',
        username: 'cdcdcd',
        password: 'cdcdcd',
      })
      .expect(200);

    assert.equal(true, true);

    Application.container.restore('/Services/Git');
  });
});
```

**Reply by thetutlage** [reply]
How are you using the `Services/Git` in your controller? Can you share that too?

**Reply by 4konung** [reply]
I've tried what you recommended but nothing changed. The real Twilio service calls his 'sendSMS' method during the tests.
Maybe I misunderstood how to use Providers?
[Here](https://github.com/4konung/adonisjs-v5-fake-service) is example how I use.

**Reply by eidellev** [reply]
Thanks @thetutlage ,

This is how I'm registering my service:
```typescript
// providers/AppProvider.ts
import { IocContract } from '@adonisjs/fold';
import GitService from 'App/Services/GitService';

export default class AppProvider {
  constructor(protected $container: IocContract) {}

  public async register() {
    // Register your own bindings
    const Logger = (await import('@ioc:Adonis/Core/Logger')).default;
    this.$container.singleton('Services/Git', () => new GitService(Logger));
  }

  public boot() {
    // IoC container is ready
  }

  public shutdown() {
    // Cleanup, since app is going down
  }

  public ready() {
    // App is ready
  }
}

```

And this is my controller:
```typescript
// app/Controllers/Http/GitController.ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext';
import { schema, rules } from '@ioc:Adonis/Core/Validator';
import GitService from '@ioc:Services/Git';

const gitCredentialsSchema = {
  url: schema.string({}, [rules.httpsUrl()]),
  username: schema.string(),
  password: schema.string(),
};

export default class GitController {
  public async checkout({ request, params }: HttpContextContract) {
    const { branch }: { branch: string | undefined } = params;

    const { url, username, password } = await request.validate({
      schema: schema.create(gitCredentialsSchema),
    });

    return GitService.checkout(url, username, password, branch);
  }

  public async createBranch({ request }: HttpContextContract) {
    const { url, username, password, fromBranch, branch } = await request.validate({
      schema: schema.create({
        ...gitCredentialsSchema,
        fromBranch: schema.string(),
        branch: schema.string(),
      }),
    });

    return GitService.createBranch(url, username, password, fromBranch, branch);
  }

  public async getFile({ request }: HttpContextContract) {
    const { url, username, password, branch, file } = await request.validate({
      schema: schema.create({
        ...gitCredentialsSchema,
        branch: schema.string(),
        file: schema.string(),
      }),
    });

    return GitService.getFile(url, username, password, branch, file);
  }

  public async writeFile({ request }: HttpContextContract) {
    const { url, username, password, branch, file } = await request.validate({
      schema: schema.create({
        ...gitCredentialsSchema,
        branch: schema.string(),
        file: schema.string(),
        data: schema.object().members({
          name: schema.string(),
          steps: schema.array().members(
            schema.object().members({
              stepType: schema.string(),
            }),
          ),
        }),
      }),
    });

    return GitService.writeFile(url, username, password, branch, file, request.post().data);
  }
}
```

**Comment by vowar** [comment, marked as answer]
@thetutlage Also the description for the fake method says:

> Register a fake for an existing binding. The fakes only work when [[this.useProxies]] is invoked. AdonisJs will set invoke it automatically when running tests.

Can it be an issue with the proxies? Or there should be some additional steps? While tests are running I've console logged the container and the proxiesEnabled property is false

**Reply by eidellev** [reply]
thanks @vowar ! adding `Application.container.useProxies();` fixed my issue.

**Reply by 4konung** [reply]
where I need to add this call? japaFile or in spec file?

**Reply by eidellev** [reply]
at the top of my test suite:
```
test.group('Git Controller', () => {
  Application.container.useProxies();
...
```

**Comment by 4konung** [comment]
@thetutlage The proposed answer with adding `Application.container.useProxies()` didn't helps for all my cases. Today I found some interesting behavior. Faking works If service used in the controller method, but in inline closure, fake doesn't work. In my project, some services used even in models, and I need somehow to fake them during the tests. Any suggestions?

**Reply by thetutlage** [reply]
@4konung Its hard to answer without looking at the application lifecycle. Ideally `Application.container.useProxies` should be called as early as possible. Otherwise the services resolved before the `useProxies` calls will point to the original source and not the proxied object.

If your app is straight forward to setup and follow, then feel free to add me to the project and I can suggest something

**Reply by 4konung** [reply]
[repo](https://github.com/4konung/adonisjs-v5-fake-service) with an example. A new issue appeared while I prepared this repo. Now when `Application.container.useProxies` called in test file, I've got this error - `TypeError: User_1.default.create is not a function`, but this doesn't happen in the project.
